File: Data Structures & Algorithms/reorder-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        s=head
        f=head
        while f!=None and f.next!=None:
            f=f.next.next
            if f==None:
                break
            s=s.next
        t=s.next
        s.next=None
        s=t
        prev=None
        while s!=None:
            t=s.next
            s.next=prev
            prev=s
            s=t
        l1=head
        l2=prev

        # 2->4->None
        # 8->6->None 

        while l2!=None:
                       # 2->
                       # 8->
            t1=l1.next # ->4
            l1.next=l2 # 2->8

            t2=l2.next # ->6
            l2.next=t1 # 8->4
            
            l1=t1      # 4->
            l2=t2      # 8->
            

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy=new ListNode(0);
        dummy.next=head;
        ListNode temp=dummy.next,prev=dummy;
        
        while(temp!=null){
            ListNode kth=kth(temp,k);
            if(kth==null){
                prev.next=temp;
                break;
            }
            ListNode save=kth.next;
            kth.next=null;
            
            ListNode revHead=reverse(temp);
            prev.next=revHead;
            temp.next=save;

            prev=temp;
            temp=save;
        }
        return dummy.next;
    }
    public ListNode reverse(ListNode head){
        ListNode curr=head;
        ListNode prev=null;
        while(curr!=null){
            ListNode temp=curr.next;
            curr.next=prev;
            prev=curr;
            curr=temp;
        }
        return prev;
    }
    public ListNode kth(ListNode head,int k){
        ListNode curr=head;
        while(curr!=null && k>1){
            curr=curr.next;
            k--;
            if(k==1) return curr;
        }
        return null;
    }

}
